<a href="https://colab.research.google.com/github/EdenShaveet/Disclosure-Curriculum/blob/main/Module2_NHANES_merge_subset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module Exercise: Merge, Subset & Preprocess NHANES Data (then disclose your methods!)
**Script Description:** Merges and subsets pre-pandemic demographic and body measurement data from NHANES, and provides acommpanying methods disclosure text

**Instructions:** Download two NHANES datasets from GitHub and upload them to this script in the code blocks marked with the symbols "⬅️🗂️." Run each code block sequentially to merge and subset the datasets to pre-defined characteristics. Review the methods disclosures at the end of this script and revise/add details as you see fit.

First, let's import necessary packages.

In [15]:
# Import packages
import pandas as pd
import seaborn as sns
import io

# Download NHANES Datasets from GitHub

Download pre-pandemic (2017-2020) [demographic NHANES data](https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/P_DEMO.htm) as an xlsx file from [HERE](https://github.com/EdenShaveet/Disclosure-Curriculum/blob/main/NHANES_Demo_2017_2020.xlsx).

Download pre-pandemic (2017-2020) [body measures NHANES data](https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/P_BMX.htm) as a xlsx file from [HERE](https://github.com/EdenShaveet/Disclosure-Curriculum/blob/main/NHANES_Body_2017_2020.xlsx).

# Upload & View Data

Run the following code block, select "choose files," and select your downloaded demographic dataset (NHANES_Demo_2017_2020.xlsx)

In [16]:
# Upload Demographic dataset
from google.colab import files
uploaded = files.upload()

Saving NHANES_Demo_2017_2020.xlsx to NHANES_Demo_2017_2020 (1).xlsx


**Output Explanation:** This line of code allows you to upload files from your local file system.

In [17]:
# Name demographic dataset as "df_demo"
df_demo = pd.read_excel(io.BytesIO(uploaded['NHANES_Demo_2017_2020.xlsx']))
# Return dataset preview
df_demo

seqn  sddsrvyr  ridstatr  riagendr  ridageyr  ridagemn  ridreth1  \
0      109263        66         2         1         2       NaN         5   
1      109264        66         2         2        13       NaN         1   
2      109265        66         2         1         2       NaN         3   
3      109266        66         2         2        29       NaN         5   
4      109267        66         1         2        21       NaN         2   
...       ...       ...       ...       ...       ...       ...       ...   
15555  124818        66         2         1        40       NaN         4   
15556  124819        66         2         1         2       NaN         4   
15557  124820        66         2         2         7       NaN         3   
15558  124821        66         2         1        63       NaN         4   
15559  124822        66         2         1        74       NaN         2   

       ridreth3  ridexmon  dmdborn4  ...  fiaintrp  mialang  miaproxy  \
0             6       2.0         1  ...       2.0      NaN       NaN   
1             1       2.0         1  ...       2.0      1.0       2.0   
2             3       2.0         1  ...       2.0      NaN       NaN   
3             6       2.0         2  ...       2.0      1.0       2.0   
4             2       NaN         2  ...       2.0      NaN       NaN   
...         ...       ...       ...  ...       ...      ...       ...   
15555         4       1.0         1  ...       2.0      1.0       2.0   
15556         4       2.0         1  ...       2.0      NaN       NaN   
15557         3       2.0         1  ...       2.0      NaN       NaN   
15558         4       1.0         1  ...       2.0      1.0       2.0   
15559         2       2.0         2  ...       NaN      2.0       2.0   

       miaintrp  aialanga      wtintprp      wtmecprp  sdmvpsu  sdmvstra  \
0           NaN       NaN   7891.762435   8951.815567        3       156   
1           2.0       1.0  11689.747264  12271.157043        1       155   
2           NaN       NaN  16273.825939  16658.764203        1       157   
3           2.0       1.0   7825.646112   8154.968193        2       168   
4           NaN       NaN  26379.991724      0.000000        1       156   
...         ...       ...           ...           ...      ...       ...   
15555       2.0       1.0  21586.596728  21666.889837        1       166   
15556       NaN       NaN   1664.919253   1838.169709        2       171   
15557       NaN       NaN  14819.783161  16497.806674        1       157   
15558       2.0       1.0   4666.817952   4853.430230        1       158   
15559       2.0       NaN  14024.476634  15195.796658        1       149   

       indfmpir  
0          4.66  
1          0.83  
2          3.06  
3          5.00  
4          5.00  
...         ...  
15555      3.82  
15556      0.07  
15557      1.22  
15558      3.71  
15559       NaN  

[15560 rows x 29 columns]

Let's take a look at the shape of our diet dataset

*Based on information from NHANES, we expect to see **15,560** cases (rows) and **29** variables (columns)*

In [18]:
df_demo.shape

(15560, 29)

**Output Explanation:** This output shows you the *shape* of your dataset as (# rows, # columns)

Run the following code block, select "choose files," and select your downloaded body measures dataset (NHANES_Body_2017_2020.xlsx)

In [19]:
# Upload Body Measures dataset
from google.colab import files
uploaded = files.upload()

Saving NHANES_Body_2017_2020.xlsx to NHANES_Body_2017_2020 (1).xlsx


**Output Explanation:** This line of code allows you to upload files from your local file system.

In [20]:
# Name diet dataset as "df_body"
df_body = pd.read_excel(io.BytesIO(uploaded['NHANES_Body_2017_2020.xlsx']))
# Return dataset preview
df_body

seqn  bmdstats  bmxwt  bmiwt  bmxrecum  bmirecum  bmxhead  bmihead  \
0      109263         4    NaN    NaN       NaN       NaN      NaN      NaN   
1      109264         1   42.2    NaN       NaN       NaN      NaN      NaN   
2      109265         1   12.0    NaN      91.6       NaN      NaN      NaN   
3      109266         1   97.1    NaN       NaN       NaN      NaN      NaN   
4      109269         3   13.6    NaN      90.9       NaN      NaN      NaN   
...       ...       ...    ...    ...       ...       ...      ...      ...   
14295  124818         1  108.8    NaN       NaN       NaN      NaN      NaN   
14296  124819         1   15.4    NaN      94.9       NaN      NaN      NaN   
14297  124820         1   22.9    NaN       NaN       NaN      NaN      NaN   
14298  124821         1   79.5    NaN       NaN       NaN      NaN      NaN   
14299  124822         1   59.7    NaN       NaN       NaN      NaN      NaN   

       bmxht  bmiht  ...  bmxleg  bmileg  bmxarml  bmiarml  bmxarmc  bmiarmc  \
0        NaN    NaN  ...     NaN     NaN      NaN      NaN      NaN      NaN   
1      154.7    NaN  ...    36.3     NaN     33.8      NaN     22.7      NaN   
2       89.3    NaN  ...     NaN     NaN     18.6      NaN     14.8      NaN   
3      160.2    NaN  ...    40.8     NaN     34.7      NaN     35.8      NaN   
4        NaN    1.0  ...     NaN     NaN      NaN      1.0      NaN      1.0   
...      ...    ...  ...     ...     ...      ...      ...      ...      ...   
14295  168.7    3.0  ...    45.6     NaN     38.6      NaN     33.6      NaN   
14296   93.7    NaN  ...     NaN     NaN     17.5      NaN     15.8      NaN   
14297  123.3    NaN  ...     NaN     NaN     24.3      NaN     18.5      NaN   
14298  176.4    NaN  ...    42.0     NaN     39.5      NaN     31.4      NaN   
14299  167.5    NaN  ...    41.1     NaN     40.3      NaN     29.2      NaN   

       bmxwaist  bmiwaist  bmxhip  bmihip  
0           NaN       NaN     NaN     NaN  
1          63.8       NaN    85.0     NaN  
2          41.2       NaN     NaN     NaN  
3         117.9       NaN   126.1     NaN  
4           NaN       1.0     NaN     NaN  
...         ...       ...     ...     ...  
14295     114.7       NaN   118.0     NaN  
14296      48.4       NaN     NaN     NaN  
14297      57.5       NaN     NaN     NaN  
14298      97.1       NaN    99.8     NaN  
14299      86.9       NaN    90.5     NaN  

[14300 rows x 22 columns]

Let's take a look at the shape of our diet dataset.

*Based on information from NHANES, we expect to see **14,300** cases or rows and **22** columns or variables*

In [21]:
df_body.shape

(14300, 22)

**Output Explanation:** This output shows you the *shape* of your dataset as (# rows, # columns)

# Merge Datasets

We are going to merge our two datasets on the unique NHANES respondent identifier: "SEQN"

In [22]:
# Merge demographic and body measures datasets
df_merged = pd.merge(df_demo, df_body, on='seqn', how ='inner')
# Return merged dataset
df_merged

seqn  sddsrvyr  ridstatr  riagendr  ridageyr  ridagemn  ridreth1  \
0      109263        66         2         1         2       NaN         5   
1      109264        66         2         2        13       NaN         1   
2      109265        66         2         1         2       NaN         3   
3      109266        66         2         2        29       NaN         5   
4      109269        66         2         1         2       NaN         2   
...       ...       ...       ...       ...       ...       ...       ...   
14295  124818        66         2         1        40       NaN         4   
14296  124819        66         2         1         2       NaN         4   
14297  124820        66         2         2         7       NaN         3   
14298  124821        66         2         1        63       NaN         4   
14299  124822        66         2         1        74       NaN         2   

       ridreth3  ridexmon  dmdborn4  ...  bmxleg  bmileg  bmxarml  bmiarml  \
0             6       2.0         1  ...     NaN     NaN      NaN      NaN   
1             1       2.0         1  ...    36.3     NaN     33.8      NaN   
2             3       2.0         1  ...     NaN     NaN     18.6      NaN   
3             6       2.0         2  ...    40.8     NaN     34.7      NaN   
4             2       1.0         1  ...     NaN     NaN      NaN      1.0   
...         ...       ...       ...  ...     ...     ...      ...      ...   
14295         4       1.0         1  ...    45.6     NaN     38.6      NaN   
14296         4       2.0         1  ...     NaN     NaN     17.5      NaN   
14297         3       2.0         1  ...     NaN     NaN     24.3      NaN   
14298         4       1.0         1  ...    42.0     NaN     39.5      NaN   
14299         2       2.0         2  ...    41.1     NaN     40.3      NaN   

       bmxarmc  bmiarmc  bmxwaist  bmiwaist  bmxhip  bmihip  
0          NaN      NaN       NaN       NaN     NaN     NaN  
1         22.7      NaN      63.8       NaN    85.0     NaN  
2         14.8      NaN      41.2       NaN     NaN     NaN  
3         35.8      NaN     117.9       NaN   126.1     NaN  
4          NaN      1.0       NaN       1.0     NaN     NaN  
...        ...      ...       ...       ...     ...     ...  
14295     33.6      NaN     114.7       NaN   118.0     NaN  
14296     15.8      NaN      48.4       NaN     NaN     NaN  
14297     18.5      NaN      57.5       NaN     NaN     NaN  
14298     31.4      NaN      97.1       NaN    99.8     NaN  
14299     29.2      NaN      86.9       NaN    90.5     NaN  

[14300 rows x 50 columns]

Let's take a look at the shape of our merged dataset.

*Since we performed an inner merge, we expect to see **14,300** cases (the same number as our least populated dataset: our body measures dataset) and **50** variables because our demographic dataset contains 29 variables, our dietary dataset contains 22 variables, and we merged on one variable. (29+22)-1=50*

In [23]:
# View dataset shape (rows, columns)
df_merged.shape

(14300, 50)

**Output Explanation:** This output shows you the *shape* of your dataset as (# rows, # columns)

# Subset Dataset (Exclusions)

Let's say we are only interested in maintaining a dataset that contains the demographic and body measurement data of females aged 18+ years who were born in a U.S. state or Washington D.C.

We're going to subset our dataset to include only those individuals.

In [24]:
# Subset to females 18+ born in U.S. state or D.C.
df = df_merged[(df_merged.riagendr==2) & (df_merged.ridageyr>17) & (df_merged.dmdborn4==1)]
df

seqn  sddsrvyr  ridstatr  riagendr  ridageyr  ridagemn  ridreth1  \
24     109290        66         2         2        68       NaN         4   
25     109291        66         2         2        42       NaN         5   
31     109297        66         2         2        30       NaN         5   
44     109312        66         2         2        48       NaN         3   
47     109315        66         2         2        30       NaN         4   
...       ...       ...       ...       ...       ...       ...       ...   
14274  124796        66         2         2        61       NaN         4   
14277  124799        66         2         2        80       NaN         3   
14286  124809        66         2         2        23       NaN         1   
14287  124810        66         2         2        56       NaN         4   
14289  124812        66         2         2        62       NaN         2   

       ridreth3  ridexmon  dmdborn4  ...  bmxleg  bmileg  bmxarml  bmiarml  \
24            4       2.0         1  ...    38.0     NaN     37.4      NaN   
25            6       1.0         1  ...     NaN     1.0      NaN      1.0   
31            6       1.0         1  ...    34.0     NaN     34.6      NaN   
44            3       2.0         1  ...    37.6     NaN     37.5      NaN   
47            4       1.0         1  ...    37.6     NaN     35.4      NaN   
...         ...       ...       ...  ...     ...     ...      ...      ...   
14274         4       2.0         1  ...    36.5     NaN     38.4      NaN   
14277         3       1.0         1  ...    34.7     NaN     35.2      NaN   
14286         1       1.0         1  ...    38.4     NaN     32.6      NaN   
14287         4       1.0         1  ...    43.0     NaN     42.2      NaN   
14289         2       1.0         1  ...    37.0     NaN     36.2      NaN   

       bmxarmc  bmiarmc  bmxwaist  bmiwaist  bmxhip  bmihip  
24        31.8      NaN      92.0       NaN   106.2     NaN  
25         NaN      1.0       NaN       1.0     NaN     1.0  
31        28.3      NaN      73.2       NaN    92.5     NaN  
44        32.0      NaN      95.7       NaN    96.3     NaN  
47        32.7      NaN      98.7       NaN   107.7     NaN  
...        ...      ...       ...       ...     ...     ...  
14274     35.7      NaN     115.7       NaN   120.2     NaN  
14277     27.2      NaN      77.5       NaN    99.5     NaN  
14286     30.7      NaN      92.5       NaN   103.8     NaN  
14287     40.7      NaN     120.0       NaN   128.4     NaN  
14289     31.4      NaN      99.3       NaN   104.6     NaN  

[3258 rows x 50 columns]

Let's say we also wish to exclude any case that is missing measurements for upper arm length or arm circumference

In [25]:
# Exclude those who did not provide measurements for upper arm length or arm circumference.
df = df[(df.bmiarml != 1) | (df.bmiarmc != 1)]
df

seqn  sddsrvyr  ridstatr  riagendr  ridageyr  ridagemn  ridreth1  \
24     109290        66         2         2        68       NaN         4   
31     109297        66         2         2        30       NaN         5   
44     109312        66         2         2        48       NaN         3   
47     109315        66         2         2        30       NaN         4   
48     109316        66         2         2        62       NaN         4   
...       ...       ...       ...       ...       ...       ...       ...   
14274  124796        66         2         2        61       NaN         4   
14277  124799        66         2         2        80       NaN         3   
14286  124809        66         2         2        23       NaN         1   
14287  124810        66         2         2        56       NaN         4   
14289  124812        66         2         2        62       NaN         2   

       ridreth3  ridexmon  dmdborn4  ...  bmxleg  bmileg  bmxarml  bmiarml  \
24            4       2.0         1  ...    38.0     NaN     37.4      NaN   
31            6       1.0         1  ...    34.0     NaN     34.6      NaN   
44            3       2.0         1  ...    37.6     NaN     37.5      NaN   
47            4       1.0         1  ...    37.6     NaN     35.4      NaN   
48            4       1.0         1  ...    40.5     NaN     38.5      NaN   
...         ...       ...       ...  ...     ...     ...      ...      ...   
14274         4       2.0         1  ...    36.5     NaN     38.4      NaN   
14277         3       1.0         1  ...    34.7     NaN     35.2      NaN   
14286         1       1.0         1  ...    38.4     NaN     32.6      NaN   
14287         4       1.0         1  ...    43.0     NaN     42.2      NaN   
14289         2       1.0         1  ...    37.0     NaN     36.2      NaN   

       bmxarmc  bmiarmc  bmxwaist  bmiwaist  bmxhip  bmihip  
24        31.8      NaN      92.0       NaN   106.2     NaN  
31        28.3      NaN      73.2       NaN    92.5     NaN  
44        32.0      NaN      95.7       NaN    96.3     NaN  
47        32.7      NaN      98.7       NaN   107.7     NaN  
48        35.7      NaN      97.5       NaN   102.0     NaN  
...        ...      ...       ...       ...     ...     ...  
14274     35.7      NaN     115.7       NaN   120.2     NaN  
14277     27.2      NaN      77.5       NaN    99.5     NaN  
14286     30.7      NaN      92.5       NaN   103.8     NaN  
14287     40.7      NaN     120.0       NaN   128.4     NaN  
14289     31.4      NaN      99.3       NaN   104.6     NaN  

[3154 rows x 50 columns]

# Download Dataset

Let's download our new dataset and save it in a place we'll remember (You'll need it again in a later module)

In [26]:
df.to_csv('NHANES_subset.csv')
files.download('NHANES_subset.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Output Explanation:** These lines of code allow you to download our preprocessed dataset to your local device as "NHANES_subset.csv"

# Methods Disclosure

Now, let's **disclose** our data acquisition and preprocessing (merging, subsetting/excluding) methods in such a way that if someone wanted to reproduce our methods, they could.

*Feel free to add your own revisions or additions to this disclosure!*

1. **Data Acquisition** (Disclose ***how*** you got your data. )
* These data were acquired online in May 2022 from two pre-pandemic NHANES datasets: [2017-2020 Demographic Variables and Sample Weights](https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/P_DEMO.htm#RIDAGEYR) and [2017-2020 Body Measures](https://wwwn.cdc.gov/Nchs/Nhanes/2017-2018/P_BMX.htm#BMIWAIST). Information about sample eligibility is available in each set's linked documentation. Information about NHANES data collection procedures and survey operations is available in the [NHANES Overview Documentation](https://www.cdc.gov/nchs/data/nhanes/nhanes_13_14/NHANES_Overview_Brochure.pdf). Each dataset was converted from an XPT file to a XLSX file using Stata v.17 and made available in a [GitHub repository](https://github.com/EdenShaveet/Disclosure-Curriculum).

2. **Data Merging** (Disclose if your data were ***combined*** with other data and provide code to the extent possible)
* An inner merge on participant identifier number (seqn) was conducted using the Pandas package in Python within a Google Colab notebook. See [Colab notebook](https://colab.research.google.com/drive/1gcERSr_4ySrOf_ow1Wtpw6uLSAQpBOQo?usp=sharing) for code.

3. **Data Subsetting** (Disclose any data ***preprocessing*** approaches and provide code to the extent possible)
* The merged dataset was subset to include only females aged 18+ at the time of participation who were born in a U.S. state or D.C. We additionally excluded respondents for whom upper arm length or arm circumference measurements were not obtained. See [Colab notebook](https://colab.research.google.com/drive/1gcERSr_4ySrOf_ow1Wtpw6uLSAQpBOQo?usp=sharing) for code.
